pkg/controller/operators/olm/operator.go

```diff
@@ -1369,3 +1374,43 @@ func (a *Operator) ensureDeploymentAnnotations(logger *logrus.Entry, csv *v1alph
 	}
 	return utilerrors.NewAggregate(updateErrs)
 }
+
+// Clean up orphaned deployments after reinstalling deployments process
+func (a *Operator) cleanupOrphanedDeployments(logger *logrus.Entry, csv *v1alpha1.ClusterServiceVersion) error {
+	// Extract the InstallStrategy for the deployment
+	strategy, err := a.resolver.UnmarshalStrategy(csv.Spec.InstallStrategy)
+	if err != nil {
+		logger.Warn("could not parse install strategy while cleaning up CSV deployment")
+		return nil
+	}
+
+	// Assume the strategy is for a deployment
+	strategyDetailsDeployment, ok := strategy.(*install.StrategyDetailsDeployment)
+	if !ok {
+		logger.Warnf("could not cast install strategy as type %T", strategyDetailsDeployment)
+		return nil
+	}
+
+	depNames := map[string]string{}
+	for _, dep := range strategyDetailsDeployment.DeploymentSpecs {
+		depNames[dep.Name] = dep.Name
+	}
+	// Get existing deployments in CSV's namespace and owned by CSV
+	existingDeployments, err := a.lister.AppsV1().DeploymentLister().Deployments(csv.GetNamespace()).List(ownerutil.CSVOwnerSelector(csv))
+	if err != nil {
+		return err
+	}
+
+	// compare existing deployments to deployments in CSV's spec to see if any need to be deleted
+	for _, d := range existingDeployments {
+		if _, exists := depNames[d.GetName()]; !exists {
+			logger.Infof("found an orphaned deployment %s in namespace %s", d.GetName(), csv.GetNamespace())
+			if err := a.OpClient.DeleteDeployment(csv.GetNamespace(), d.GetName(), &metav1.DeleteOptions{}); err != nil {
```

Comment:
The owner check has already been done in line #1399 above:

https://github.com/operator-framework/operator-lifecycle-manager/pull/759/files#diff-c09af0ae4ca3b7464670dbd80be66c70R1399

The `existingDeployments` should only contain deployments that belong to to a specific CSV `out` based on the `labels.selector` filter.

Reply:
That does query by the labels which we use for finding the objects, but for these objects (since they're in the same namespace) we add "real" owner references as well. It's better to be extra safe when GCing